packages/typespec-client-generator-core/src/package.ts

```diff
@@ -419,35 +420,35 @@ function getSdkEndpointParameter(
 ): [SdkEndpointParameter, readonly Diagnostic[]] {
   const diagnostics = createDiagnosticCollector();
   const servers = getServers(context.program, client.service);
-  let type: SdkEndpointType;
+  let type: SdkEndpointType | SdkUnionType;
   let optional: boolean = false;
+  const defaultOverridableEndpointType: SdkEndpointType = {
+    kind: "endpoint",
+    serverUrl: "{endpoint}",
+    templateArguments: [
+      {
+        name: "endpoint",
+        isGeneratedName: true,
+        description: "Service host",
+        kind: "path",
+        onClient: true,
+        urlEncode: false,
+        optional: false,
+        serializedName: "endpoint",
+        correspondingMethodParams: [],
+        type: getTypeSpecBuiltInType(context, "string"),
+        isApiVersionParam: false,
+        apiVersions: context.__tspTypeToApiVersions.get(client.type)!,
+        crossLanguageDefinitionId: `${getCrossLanguageDefinitionId(context, client.service)}.endpoint`,
+        decorators: [],
+      },
+    ],
+    decorators: [],
+  };
   if (servers === undefined || servers.length > 1) {
     // if there is no defined server url, or if there are more than one
     // we will return a mandatory endpoint parameter in initialization
-    const name = "endpoint";
-    type = {
-      kind: "endpoint",
-      serverUrl: "{endpoint}",
-      templateArguments: [
-        {
-          name,
-          isGeneratedName: true,
-          description: "Service host",
-          kind: "path",
-          onClient: true,
-          urlEncode: false,
-          optional: false,
-          serializedName: "endpoint",
-          correspondingMethodParams: [],
-          type: getTypeSpecBuiltInType(context, "string"),
-          isApiVersionParam: false,
-          apiVersions: context.__tspTypeToApiVersions.get(client.type)!,
-          crossLanguageDefinitionId: `${getCrossLanguageDefinitionId(context, client.service)}.endpoint`,
-          decorators: [],
-        },
-      ],
-      decorators: [],
-    };
+    type = defaultOverridableEndpointType;
```

Comment:
since we decided to use union as the endpoint type, for multi server url, why not also use union to generate all the defined endpoint?